# 检查跳跃关联度

## 导入模块

In [1]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import feather

## 读入数据

### 日线

In [2]:
price_1d = feather.read_dataframe('../data/StockPriceK1d_20241231.feather')
price_1d = price_1d[(price_1d['date'] >= '2019-01-01') & (price_1d['date'] <= '2024-12-31')]
price_1d = price_1d[(price_1d['ret'] > 0.01) | (price_1d['ret'] < -0.01)]

### 跳跃

In [3]:
jump = feather.read_dataframe('../data/jump/jump.feather')
jump = pd.merge(
    jump,
    price_1d[['issue', 'date']],
    on=['issue', 'date'],
    how='inner'
)

### 调仓日和对应的构建关联度开始时间 (120 天前)

In [8]:
trade_date = price_1d['date'].sort_values().unique()
mes = pd.date_range(start='2019-01-01', end='2024-12-31', freq='1ME')
adj_date = np.array([], dtype=np.datetime64)
calc_start_date = np.array([], dtype=np.datetime64)

for me in mes:
    trade_date_before = trade_date[trade_date <= me]
    ad = trade_date_before[-1]
    adj_date = np.append(adj_date, ad)
    if (len(trade_date_before) > 120):
        csd = trade_date_before[-120]
    else:
        csd = trade_date_before[0]
    calc_start_date = np.append(calc_start_date, csd)

In [10]:
ad = adj_date[7]
ad_str = ad.strftime('%Y%m%d')
csd = calc_start_date[7]
print(ad, csd)

2019-08-30 00:00:00 2019-03-11 00:00:00


### 跳跃关联度

In [11]:
corr_num = feather.read_dataframe(f'../data/corr/corr_num_{ad_str}.feather')
corr_size = feather.read_dataframe(f'../data/corr/corr_size_{ad_str}.feather')

## 检查 601921, 300556, 301486 在 2024-08-20 的关联跳跃

In [12]:
display(jump[(jump['date'] == '2024-08-20') & (jump['issue'] == '601921')])
display(jump[(jump['date'] == '2024-08-20') & (jump['issue'] == '300556')])
display(jump[(jump['date'] == '2024-08-20') & (jump['issue'] == '301486')])

issue       date  jump  jump_count  ret_jump  ret_nojump  \
1473279  601921 2024-08-20  True           2  0.077015    0.018295   

         ret_posjump  ret_negjump  year_mon  
1473279     0.077015          0.0    202408

issue       date  jump  jump_count  ret_jump  ret_nojump  \
857894  300556 2024-08-20  True           2  0.067435    0.010274   

        ret_posjump  ret_negjump  year_mon  
857894     0.067435          0.0    202408

issue       date  jump  jump_count  ret_jump  ret_nojump  \
1120096  301486 2024-08-20  True           2  0.077936   -0.007949   

         ret_posjump  ret_negjump  year_mon  
1120096     0.077936          0.0    202408

## 检查股票 000001 与 000002 在 2019-08-30 的跳跃关联度

In [13]:
issue_i = '000001'
issue_j = '000002'
idx_date = (jump['date'] >= csd) & (jump['date'] <= ad)

### 股票 i, j 的跳跃

In [14]:
jump_i = jump[idx_date & (jump['issue'] == issue_i)].copy()
jump_i

issue       date  jump  jump_count  ret_jump    ret_nojump  ret_posjump  \
11  000001 2019-03-25  True           1 -0.015206 -2.366492e-02     0.000000   
12  000001 2019-03-27  True           1  0.011504  1.137299e-02     0.011504   
13  000001 2019-03-28  True           1 -0.005703 -7.305650e-03     0.000000   
14  000001 2019-03-29  True           2  0.020121  2.781167e-02     0.020121   
15  000001 2019-04-01  True           1  0.020062  7.631676e-03     0.020062   
16  000001 2019-04-02  True           1  0.007559  6.006024e-03     0.007559   
17  000001 2019-04-11  True           1  0.015896 -2.983124e-02     0.015896   
18  000001 2019-04-15  True           2  0.040882 -2.096210e-02     0.040882   
19  000001 2019-04-17  True           1 -0.011035 -4.866190e-03     0.000000   
20  000001 2019-04-24  True           1  0.026650 -6.922811e-04     0.026650   
21  000001 2019-04-30  True           1 -0.010874 -7.015953e-03     0.000000   
22  000001 2019-05-06  True           1 -0.055673 -1.771321e-02     0.000000   
23  000001 2019-05-08  True           1 -0.017920 -9.478744e-03     0.000000   
24  000001 2019-05-10  True           1  0.014694  2.717993e-02     0.014694   
25  000001 2019-05-13  True           1 -0.027991 -2.436055e-03     0.000000   
26  000001 2019-05-14  True           1  0.015026  3.030772e-04     0.015026   
27  000001 2019-05-15  True           1  0.010097  2.375100e-02     0.010097   
28  000001 2019-05-21  True           2  0.014314  1.207174e-04     0.014314   
29  000001 2019-05-22  True           1 -0.004785 -8.035990e-03     0.000000   
30  000001 2019-05-30  True           1 -0.002464 -1.215845e-02     0.000000   
31  000001 2019-06-05  True           1  0.010076 -3.838076e-16     0.010076   
32  000001 2019-06-11  True           1  0.008685  1.612579e-02     0.008685   
33  000001 2019-06-18  True           1  0.007084  3.124322e-03     0.007084   
34  000001 2019-06-19  True           2  0.024693 -3.818256e-03     0.037567   
35  000001 2019-06-27  True           3  0.022107  3.005038e-03     0.022107   
36  000001 2019-07-01  True           1  0.016553 -5.726573e-03     0.016553   
37  000001 2019-07-02  True           1  0.012173  5.614293e-03     0.012173   
38  000001 2019-07-03  True           1 -0.010635 -1.426534e-03     0.000000   
39  000001 2019-08-02  True           1 -0.023682 -2.181026e-03     0.000000   
40  000001 2019-08-08  True           3  0.041018  1.917244e-02     0.041018   
41  000001 2019-08-12  True           2  0.019627  2.086479e-02     0.019627   
42  000001 2019-08-13  True           1 -0.007968 -7.360354e-03     0.000000   
43  000001 2019-08-21  True          23 -0.036335 -3.540278e-04     0.002734   

    ret_negjump  year_mon  
11    -0.015206    201903  
12     0.000000    201903  
13    -0.005703    201903  
14     0.000000    201903  
15     0.000000    201904  
16     0.000000    201904  
17     0.000000    201904  
18     0.000000    201904  
19    -0.011035    201904  
20     0.000000    201904  
21    -0.010874    201904  
22    -0.055673    201905  
23    -0.017920    201905  
24     0.000000    201905  
25    -0.027991    201905  
26     0.000000    201905  
27     0.000000    201905  
28     0.000000    201905  
29    -0.004785    201905  
30    -0.002464    201905  
31     0.000000    201906  
32     0.000000    201906  
33     0.000000    201906  
34    -0.012874    201906  
35     0.000000    201906  
36     0.000000    201907  
37     0.000000    201907  
38    -0.010635    201907  
39    -0.023682    201908  
40     0.000000    201908  
41     0.000000    201908  
42    -0.007968    201908  
43    -0.039069    201908

In [15]:
jump_j = jump[idx_date & (jump['issue'] == issue_j)].copy()
jump_j

issue       date  jump  jump_count  ret_jump  ret_nojump  ret_posjump  \
418  000002 2019-03-18  True           4  0.024030    0.021242     0.024030   
419  000002 2019-03-20  True           1  0.015409   -0.002457     0.015409   
420  000002 2019-03-25  True           1 -0.012063   -0.023861     0.000000   
421  000002 2019-03-29  True           3  0.022184    0.042354     0.022184   
422  000002 2019-04-09  True           2  0.026902    0.032790     0.026902   
423  000002 2019-04-10  True           1 -0.017476   -0.003959     0.000000   
424  000002 2019-04-22  True           5 -0.035522   -0.029101     0.000000   
425  000002 2019-04-26  True           1 -0.009916   -0.000688     0.000000   
426  000002 2019-04-30  True           2 -0.023809    0.007666     0.000000   
427  000002 2019-05-06  True           1 -0.030945   -0.032671     0.000000   
428  000002 2019-05-07  True           2  0.022978    0.008979     0.022978   
429  000002 2019-05-08  True           3 -0.016940   -0.004372     0.013763   
430  000002 2019-05-14  True           1 -0.014001   -0.002172     0.000000   
431  000002 2019-05-15  True           1  0.009017    0.006086     0.009017   
432  000002 2019-05-17  True           2 -0.009003   -0.018752     0.000000   
433  000002 2019-05-23  True           1 -0.007751   -0.015919     0.000000   
434  000002 2019-05-28  True           1  0.014956    0.007748     0.014956   
435  000002 2019-05-29  True           1 -0.011653   -0.005141     0.000000   
436  000002 2019-06-05  True           2  0.018758    0.008621     0.018758   
437  000002 2019-06-11  True           1  0.011416    0.007109     0.011416   
438  000002 2019-07-01  True           1  0.013217    0.015493     0.013217   
439  000002 2019-07-04  True           1  0.014271   -0.001328     0.014271   
440  000002 2019-07-19  True           1  0.020070    0.000326     0.020070   
441  000002 2019-07-23  True           2 -0.015435   -0.020575     0.000000   
442  000002 2019-07-31  True          16 -0.013221   -0.022972     0.002432   
443  000002 2019-08-01  True           1 -0.009081   -0.002811     0.000000   
444  000002 2019-08-02  True           1 -0.023857    0.009326     0.000000   
445  000002 2019-08-05  True           2 -0.019467   -0.016887     0.000000   
446  000002 2019-08-12  True           2  0.011271    0.006165     0.011271   
447  000002 2019-08-13  True           2 -0.009037   -0.006935     0.000000   
448  000002 2019-08-19  True           2  0.014312    0.008262     0.023318   
449  000002 2019-08-20  True           2  0.011451   -0.000720     0.011451   
450  000002 2019-08-26  True           1 -0.018720   -0.000378     0.000000   

     ret_negjump  year_mon  
418     0.000000    201903  
419     0.000000    201903  
420    -0.012063    201903  
421     0.000000    201903  
422     0.000000    201904  
423    -0.017476    201904  
424    -0.035522    201904  
425    -0.009916    201904  
426    -0.023809    201904  
427    -0.030945    201905  
428     0.000000    201905  
429    -0.030704    201905  
430    -0.014001    201905  
431     0.000000    201905  
432    -0.009003    201905  
433    -0.007751    201905  
434     0.000000    201905  
435    -0.011653    201905  
436     0.000000    201906  
437     0.000000    201906  
438     0.000000    201907  
439     0.000000    201907  
440     0.000000    201907  
441    -0.015435    201907  
442    -0.015653    201907  
443    -0.009081    201908  
444    -0.023857    201908  
445    -0.019467    201908  
446     0.000000    201908  
447    -0.009037    201908  
448    -0.009006    201908  
449     0.000000    201908  
450    -0.018720    201908

### 以股票 i 为焦点的关联跳跃

| jump_date_i | direction | corr_jump_date_j |
| :---------: | :-------: | :--------------: |
| 3.29        | +         | 3.29             |
| 4.1         | +         |                  |
| 4.2         | -         |                  |
| 4.11        | +         | 4.11             |
| 4.15        | +         |                  |
| 4.30        | -         | 4.30             |
| 5.9         | +         | 5.8              |
| 5.10        | -         | 5.10             |
| 5.14        | +         |                  |
| 5.15        | +         |                  |
| 5.17        | -         | 5.17             |
| 5.22        | -         |                  |
| 5.30        | -         |                  |
| 6.11        | +         | 6.11             |
| 6.18        | +         |                  |
| 6.19        | -         |                  |
| 6.21        | -         |                  |
| 6.27        | +         |                  |
| 7.2         | +         | 7.3              |
| 7.3         | -         |                  |
| 7.16        | -         |                  |
| 7.22        | -         | 7.23             |
| 7.31        | -         | 7.31             |
| 8.2         | -         | 8.2              |
| 8.5         | -         | 8.5              |
| 8.7         | +         | 8.7              |
| 8.8         | +         | 8.8              |
| 8.12        | +         | 8.13             |
| 8.13        | +         | 8.13             |
| 8.21        | -         |                  |
| 8.23        | -         |                  |
| 8.26        | -         |                  |

In [10]:
print(corr_num.loc[issue_i, issue_j], corr_size.loc[issue_i, issue_j])

0.5625 0.5936543984419171


In [11]:
dates = np.array([
    '2019-03-29', '2019-04-01', '2019-04-02', '2019-04-11', '2019-04-15',
    '2019-04-30', '2019-05-09', '2019-05-10', '2019-05-14', '2019-05-15',
    '2019-05-17', '2019-05-22', '2019-05-30', '2019-06-11', '2019-06-18',
    '2019-06-19', '2019-06-21', '2019-06-27', '2019-07-02', '2019-07-03',
    '2019-07-16', '2019-07-22', '2019-07-31', '2019-08-02', '2019-08-05',
    '2019-08-07', '2019-08-08', '2019-08-12', '2019-08-13', '2019-08-21',
    '2019-08-23', '2019-08-26'
], dtype=np.datetime64)
dates_corr = np.array([
    '2019-03-29', '2019-04-11', '2019-04-30', '2019-05-09', '2019-05-10',
    '2019-05-17', '2019-06-11', '2019-07-02', '2019-07-22', '2019-07-31',
    '2019-08-02', '2019-08-05', '2019-08-07', '2019-08-08', '2019-08-12',
    '2019-08-13'
], dtype=np.datetime64)

In [12]:
len(dates_corr) / len(dates)

0.5

In [13]:
jump_i['abs_ret_jump'] = jump_i['ret_jump'].apply(np.abs)
jump_i.loc[jump_i['date'].isin(dates_corr), 'abs_ret_jump'].sum() / jump_i['abs_ret_jump'].sum()

np.float64(0.524812275828415)